rust/src/main.rs

```diff
@@ -79,12 +76,15 @@ fn setup_logging() -> Result<(), fern::InitError> {
         }).level(level_filter)
         .chain(io::stderr());
 
-    if let Some(logging_file_path) = path_for_log_file(XI_LOG_FILE) {
-        fern_dispatch = fern_dispatch.chain(fern::log_file(logging_file_path)?)
-    } else {
-        eprintln!(
-            "[WARNING] There was an issue getting the path for the log file, just using stderr"
-        );
+    match path_for_log_file(XI_LOG_FILE) {
+        Ok(logging_file_path) => {
+            fern_dispatch = fern_dispatch.chain(fern::log_file(logging_file_path)?)
+        }
+        Err(e) => {
+            let error_message =
+                "[WARNING] There was an issue getting the path for the log file, falling back to stderr.";
+            eprintln!("{}\nError:\n{:?}", error_message, e)
```

Comment:
This could probably be reordered a bit so it finalizes the (stderr-only) logger and *then* prints the error message using the appropriate macro.

Reply:
Yeah I was half and half on whether to do the error logging before or after starting fern.

I've given it a go. Please tell me if there's a cleaner way to do the `apply()`s and then log the error. 🙂

Reply:
You could get creative with if let:
```
let p = path_for_log_file(XI_LOG_FILE);
if let Ok(logging_file_path) = p {
    fern_dispatch.chain(fern::log_file(logging_file_path)?);
}
fern_dispatch.apply()?;
if let Err(e) = p {
    warn!("There was an issue getting the path for the log file: {}, falling back to stderr.", e);
}
```

I think that's about as linear as you can get.

Reply:
Mmmmm. I wrote it out but idk if it looks cleaner:

Old:

```rust
let fern_dispatch = fern::Dispatch::new().etc.etc.....

match path_for_log_file(XI_LOG_FILE) {
	Ok(logging_file_path) => fern_dispatch
		.chain(fern::log_file(logging_file_path)?)
		.apply()?,
	Err(err) => {
		fern_dispatch.apply()?;
		warn!(
			"There was an issue getting the path for the log file: {}, falling back to stderr.",
			err,
		);
	}
}
info!("Logging with fern is setup");
Ok(())
```

New:

```rust
let mut fern_dispatch = fern::Dispatch::new().etc.etc.....

let p = path_for_log_file(XI_LOG_FILE);
if let Ok(logging_file_path) = &p {
	fern_dispatch = fern_dispatch.chain(fern::log_file(logging_file_path)?);
}
fern_dispatch.apply()?;
if let Err(e) = &p {
	warn!(
		"There was an issue getting the path for the log file: {:?}, falling back to stderr.",
		e,
	);
}
info!("Logging with fern is setup");
Ok(())
```
🤷‍♂️ 

I think personally I lean towards the match statement. Which do you prefer? 😃

Reply:
Yeah I get you.

I tried adding some comments and spacing and it looks much more readable than it was before & is DRY.

```rust
let mut fern_dispatch = fern::Dispatch::new().etc.etc.....

let path_result = path_for_log_file(XI_LOG_FILE);

// If the logging_file_path returned successfully, add the logfile capability to fern
if let Ok(logging_file_path) = &path_result {
	fern_dispatch = fern_dispatch.chain(fern::log_file(logging_file_path)?);
}

// Start fern
fern_dispatch.apply()?;

// If the logging_file_path returned an error, print it with fern/logs
if let Err(e) = &path_result {
	warn!(
		"There was an issue getting the path for the log file: {:?}, falling back to stderr.",
		e,
	);
}

info!("Logging with fern is setup");
Ok(())
```

Good balance of both worlds?

Reply:
I've pushed this up (with a slight change to `warn!()` to make it a bit cleaner IMO).